### PR TITLE
Training linear head script

### PR DESCRIPTION
**Pull request by piotrmwojcik**

### CODE REVIEW

File: vissl/trainer/train_task.py [hunk 1]
```diff
@@ -60,7 +60,7 @@ def __init__(self, config: AttrDict):
         self.available_splits = []
         self.base_loss = None
         self.meters = None
-        self.datasets = None
+        self.sdatasets = None
```
Comment (jano1906ideas) [inline vissl/trainer/train_task.py]: Explain please.

File: configs/config/compvits/data/test/in1k.yaml [hunk 2]
```diff
@@ -6,7 +6,7 @@ config:
       DATA_SOURCES: [disk_folder]
       LABEL_SOURCES: [disk_folder]
       DATASET_NAMES: [imagenet1k_folder]
-      DATA_PATHS: [datasets/imagenet-1k-kaggle/ILSVRC/Data/CLS-LOC/val]
+      DATA_PATHS: [/data/pwojcik/val]
```
Comment (jano1906ideas) [inline configs/config/compvits/data/test/in1k.yaml]: Revert or we have to decide on common path

File: configs/config/compvits/data/train/in1k.yaml [hunk 3]
```diff
@@ -6,7 +6,7 @@ config:
       DATA_SOURCES: [disk_folder]
       LABEL_SOURCES: [disk_folder]
       DATASET_NAMES: [imagenet1k_folder]
-      DATA_PATHS: [datasets/imagenet-1k-kaggle/ILSVRC/Data/CLS-LOC/train]
+      DATA_PATHS: [/data/pwojcik/ILSVRC/Data/CLS-LOC/train]
```
Comment (jano1906ideas) [inline configs/config/compvits/data/train/in1k.yaml]: Revert or we have to decide on common path

File: configs/config/compvits/base.yaml
Comment (jano1906ideas) [inline configs/config/compvits/base.yaml:1]: Split into 2 files: 
base_local with previous amount of resources set
base_cluster with larger batchsize, num_worksers etc.

File: compvits/utils.py
Comment (jano1906ideas) [inline compvits/utils.py:1]: I believe and pray this doesn't contain a bug

File: compvits/scripts/train_linear_head.py [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+            if re.match(targets_pattern, filename):
+                target_files.append(filename)
+
+        features = [torch.tensor(np.load(os.path.join(root_dir, f),
+                                         allow_pickle=True)) for f in features_files]
+        targets = [torch.tensor(np.load(os.path.join(root_dir, f),
+                                        allow_pickle=True)) for f in target_files]
+
+        self.features = torch.cat(features, dim=0)
+        self.targets = torch.cat(targets, dim=0)
```
Comment (jano1906ideas) [inline compvits/scripts/train_linear_head.py]: I guess this is useful if we extract features on multigpu?

File: compvits/scripts/train_linear_head.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        return self.features[idx].float(), self.targets[idx]
+
+
+class LinearClassifier(nn.Module):
+    """Linear layer to train on top of frozen features"""
+    def __init__(self, dim, num_labels=1000):
+        super(LinearClassifier, self).__init__()
+        self.num_labels = num_labels
+        self.linear = nn.Linear(dim, num_labels)
+        self.linear.weight.data.normal_(mean=0.0, std=0.01)
```
Comment (jano1906ideas) [inline compvits/scripts/train_linear_head.py:57]: Is this a standard way of initializing linear head?

File: compvits/scripts/train_linear_head.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        self.linear.bias.data.zero_()
+
+    def forward(self, x):
+        # flatten
+        x = x.view(x.size(0), -1)
```
Comment (jano1906ideas) [inline compvits/scripts/train_linear_head.py:62]: Unnecessary? We loose structure of the input, not a big deal probably, but still.
Also it doesn't provide computational benefit if data was already contiguous (and this code fails it data is not contiguous anyway)

x = torch.arange(0,9).reshape(3,3)
y = torch.permute(x, dims=(1,0))
print(x.is_contiguous()) # True
print(y.is_contiguous()) # False
y.view(9,1) # Runtime error

**Review by jano1906ideas — COMMENTED** [review]
Not finished

File: compvits/scripts/train_linear_head.py (hunk 5 above)
Comment (piotrmwojcik) [reply compvits/scripts/train_linear_head.py:57]: This is a way that this is done in DINO, let's keep it a reference

File: compvits/scripts/train_linear_head.py (hunk 6 above)
Comment (piotrmwojcik) [reply compvits/scripts/train_linear_head.py:62]: Also here we follow DINO codebase

File: compvits/scripts/test_linear.sh [hunk 7]
```diff
@@ -10,6 +10,9 @@ dir="logs/test_linear/M${M}/${model}/K$K"
 if [[ $model == "deitb" ]]; then
     head_cfg=mlp_768_1000
     trunk_cfg=deitb
+elif [[ $model == "dino" ]]; then
+    head_cfg=mlp_768_1000
```
Comment (jano1906ideas) [inline compvits/scripts/test_linear.sh:14]: mlp_768_100 -> mlp_emlp_768_1000

**Review by jano1906ideas — APPROVED** [review]
Change this one line in test_linear.sh please or rewrite the script in a way that enables choosing between mlp and eval_mlp heads.

Other than that - good ;). You can merge, please use squash and merge.

**Review by jano1906ideas — APPROVED** [review]
Ok

**Pull request merged**